server/externalscripts/rocketchat-stats

```diff
@@ -61,41 +68,87 @@ def login(api_url, username, password):
         logging.critical("Wrong data receive from API (login step)")
         logging.debug(r.json())
         return None
-    
+
     if r.json()['status'] == "success":
         logging.info("Login Success")
         return r.json()['data']
     else:
         logging.critical("Login failed, check credentials")
         return None
 
+
+def file_age_in_seconds(pathname):
+    return time.time() - os.stat(pathname)[stat.ST_MTIME]
+
+
 def statistics(api_url, username, password, statsname):
     """
     print JSON with statistics from API/v1 
     """
     statsname
     endpoint = "%s/v1/statistics" % (api_url)
-    # Login
-    credentials = login(api_url, username, password)       
-    if not credentials:
-        return None
-    
-    # Prepare headers
-    headers = { "X-Auth-Token": credentials['authToken'],
-                "X-User-Id": credentials['userId'] }
-    try:
-        r = requests.get(endpoint, headers=headers)
-    except:
-        logging.critical("Failed to connect API")
-        return None
-    
-    try:
-        r.json()
-    except:
-        logging.critical("Wrong JSON receive from API /v1/statistics, check endpoint (statistics step)")
-        return None
-    
-    if statsname in r.json()['statistics'].keys():
-        print(r.json()['statistics'][statsname])    
+
+    EXECUTABLE = os.path.basename(__file__)
+    statusfile = "/tmp/%s_%s.pickle" % (EXECUTABLE, os.getuid())
+
+    data = {}
+    reload_data = True
+    if os.path.exists(statusfile):
+        try:
+            data = pickle.load(open(statusfile, "rb"))
+        except:
+            logging.critical("ERROR: unable to read statusfile %s" % statusfile)
+        file_age = file_age_in_seconds(statusfile)
+        if file_age < max_file_age:
+            reload_data = False
+        else:
+            reload_data = True
+            os.remove(statusfile)
+
+    if reload_data:
+        # Login
+
+        if "credentials" in data:
+            # use cached auth tolen, because login to active directory/ldap take too much time
+            credentials = data["credentials"]
+        else:
+            credentials = login(api_url, username, password)
+
+        if not credentials:
+            logging.critical("login failed")
+            return None
+        # Prepare headers
+        headers = {"X-Auth-Token": credentials['authToken'],
+                   "X-User-Id": credentials['userId']}
+
+        try:
+            r = requests.get(endpoint, headers=headers)
+        except Exception as e:
+            logging.critical("Failed to connect API: %s" % e)
+            return None
+
+        try:
+            data = r.json()
+        except Exception as e:
+            logging.critical("Wrong JSON receive from API /v1/statistics, check endpoint (statistics step): %s" % e)
+            return None
+
+        try:
+            data["credentials"] = credentials
```

Comment:
```suggestion
            os.umask(0o077)
            data["credentials"] = credentials
```
Prevent authentication token file read was allowed for all users.

Reply:
fully aggree

Reply:
Great

Reply:
Dear @scoopex,
Could you integrate this suggested change to PR, it seems I can’t... (1st time using « suggested change »). Many thanks